src/stress/memory/oom.spec.ts

```diff
@@ -3,32 +3,157 @@ Stress tests covering robustness when available VRAM is exhausted.
+      buffer = device.createBuffer({
+        size,
+        usage: GPUBufferUsage.MAP_READ,
+        mappedAtCreation,
+      });
+      break;
+    case GPUMapMode.WRITE:
+      buffer = device.createBuffer({
+        size,
+        usage: GPUBufferUsage.MAP_WRITE,
+        mappedAtCreation,
+      });
+      break;
+    default:
+      unreachable();
+  }
+  // If we want the buffer to be mappable and also mappedAtCreation, we call unmap on it now.
+  if (mappable && mappedAtCreation) {
+    buffer.unmap();
+  }
+  return buffer;
+}
+
 g.test('vram_oom')
   .desc(`Tests that we can allocate buffers until we run out of VRAM.`)
   .fn(async t => {
     await exhaustVramUntilUnder64MB(t.device);
   });
 
-g.test('get_mapped_range')
-  .desc(
-    `Tests getMappedRange on a mappedAtCreation GPUBuffer that failed allocation due
-to OOM. This should throw a RangeError, but below a certain threshold may just
-crash the page.`
-  )
-  .unimplemented();
-
 g.test('map_after_vram_oom')
   .desc(
     `Allocates tons of buffers and textures with varying mapping states (unmappable,
 mappable, mapAtCreation, mapAtCreation-then-unmapped) until OOM; then attempts
-to mapAsync all the mappable objects.`
+to mapAsync all the mappable objects. The last buffer should be an error buffer so
+mapAsync on it should reject and produce a validation error. `
   )
-  .unimplemented();
+  .params(u =>
+    u
+      .combine('mappable', [true, false])
+      .combine('mode', [GPUConst.MapMode.READ, GPUConst.MapMode.WRITE])
+      .combine('mappedAtCreation', [true, false])
+      .combine('unmapBeforeResolve', [true, false])
+  )
+  .fn(async t => {
+    // Use a relatively large size to quickly hit OOM.
+    const kSize = 512 * 1024 * 1024;
+
+    const { mappable, mode, mappedAtCreation, unmapBeforeResolve } = t.params;
+    const buffers: GPUBuffer[] = [];
+    // Closure to call map and verify results on all of the buffers.
+    const finish = async () => {
+      if (mappable) {
+        await Promise.all(buffers.map(value => value.mapAsync(mode)));
+      } else {
+        buffers.forEach(value => {
+          t.expectValidationError(() => {
+            void value.mapAsync(mode);
+          });
+        });
+      }
+      // Finally, destroy all the buffers to free the resources.
+      buffers.forEach(buffer => buffer.destroy());
+    };
+
+    let errorBuffer: GPUBuffer;
+    for (;;) {
+      if (mappedAtCreation) {
+        // When mappedAtCreation is true, OOM can happen on the client which throws a RangeError. In
+        // this case, we don't do any validations on the OOM buffer.
+        try {
+          t.device.pushErrorScope('out-of-memory');
+          const buffer = createBuffer(t.device, kSize, mappable, mode, mappedAtCreation);
+          if (await t.device.popErrorScope()) {
+            errorBuffer = buffer;
+            break;
+          }
+          buffers.push(buffer);
+        } catch (ex) {
+          await finish();
```

Comment:
nit: maybe check that the exception was a RangeError

Reply:
Done.